internal/gcs/bridge.go

```diff
@@ -0,0 +1,356 @@
+	brdg.mu.Unlock()
+	brdg.log.Debug("bridge terminating")
+	brdg.conn.Close()
+}
+
+// Close closes the bridge. It is undefined to call RPC or AsyncRPC after
+// calling Close.
+func (brdg *bridge) Close() error {
+	brdg.Kill()
+	close(brdg.rpcCh)
+	return nil
+}
+
+// AsyncRPC sends an RPC request to the guest but does not wait for a response.
+// If the message cannot be sent before the context is done, then an error is
+// returned.
+func (brdg *bridge) AsyncRPC(ctx context.Context, proc rpcProc, req requestMessage, resp responseMessage) (*rpc, error) {
+	call := &rpc{
+		ch:   make(chan struct{}),
+		proc: proc,
+		req:  req,
+		resp: resp,
+	}
+	if err := ctx.Err(); err != nil {
+		return nil, err
+	}
+	// Send the request.
+	select {
+	case brdg.rpcCh <- call:
+		return call, nil
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+func (call *rpc) complete(err error) {
+	call.err = err
+	close(call.ch)
+}
+
+// Err returns the RPC's result. This may be a transport error or an error from
+// the message response.
+func (call *rpc) Err() error {
+	if call.err != nil {
+		return call.err
+	}
+	return call.resp.Err()
+}
+
+// Done returns whether the RPC has completed.
+func (call *rpc) Done() bool {
+	select {
+	case <-call.ch:
+		return true
+	default:
+		return false
+	}
+}
+
+// RPC issues a synchronous RPC request. Returns immediately if the context
+// becomes done and the message is not sent.
+//
+// If allowCancel is set and the context becomes done, returns an error without
+// waiting for a response. Avoid this on messages that are not idempotent or
+// otherwise safe to ignore the response of.
+func (brdg *bridge) RPC(ctx context.Context, proc rpcProc, req requestMessage, resp responseMessage, allowCancel bool) error {
+	call, err := brdg.AsyncRPC(ctx, proc, req, resp)
+	if err != nil {
+		return err
+	}
+	var ctxDone <-chan struct{}
+	if allowCancel {
+		// This message can be safely cancelled by ignoring the response.
+		ctxDone = ctx.Done()
+	}
+	t := time.NewTimer(brdg.Timeout)
+	defer t.Stop()
+	select {
+	case <-call.ch:
+		return call.Err()
+	case <-ctxDone:
+		brdg.log.WithField("reason", ctx.Err()).Warn("ignoring response to bridge message")
+		return ctx.Err()
+	case <-t.C:
+		brdg.log.Warn("terminating bridge due to message timeout")
+		brdg.Kill()
+		<-call.ch
+		return call.Err()
+	}
+}
+
+func (brdg *bridge) recvLoopRoutine() {
+	brdg.recvLoop()
+	// Close the connection so that the writer stops.
+	brdg.conn.Close()
+	// Fail any remaining RPCs.
+	brdg.mu.Lock()
+	rpcs := brdg.rpcs
+	brdg.rpcs = nil
+	brdg.mu.Unlock()
+	for _, call := range rpcs {
+		call.complete(errBridgeClosed)
+	}
+}
+
+func readMessage(r io.Reader) (int64, msgType, []byte, error) {
+	var h [16]byte
+	_, err := io.ReadFull(r, h[:])
+	if err != nil {
+		return 0, 0, nil, err
+	}
+	typ := msgType(binary.LittleEndian.Uint32(h[:]))
+	n := binary.LittleEndian.Uint32(h[4:])
+	id := int64(binary.LittleEndian.Uint64(h[8:]))
+	if n < 16 || n > 0x10000 {
```

Comment:
Is there a max send size? Should we make this a named constant for more clarity?